File: pt6/1233.remove-sub-folders-from-the-filesystem.go
package main

import (
	"sort"
	"strings"
)

func removeSubfolders(folder []string) []string {
	sort.Strings(folder)
	roots := []string{}

	for _, path := range folder {
		if len(roots) == 0 || !strings.HasPrefix(path, roots[len(roots)-1]+"/") {
			roots = append(roots, path)
		}
	}

	return roots
}

/*
Explanation

Sort folders lexicographically. A parent appears before all of its subfolders,
so we only need to compare each path with the latest accepted root. A path is a
subfolder exactly when it starts with root + "/".

The slash matters: "/a/b" is under "/a", but "/ab" is not. Sorting removes the
need for a trie because descendants are grouped directly after their parent.

Edge cases: sibling folders with common prefixes; deep chains like /a/b/c;
empty result is impossible because each path itself can be a root.

Time complexity: O(n log n * L) for sorting and prefix comparisons.
Space complexity: O(n) for the result.
*/

File: pt6/1233.remove-sub-folders-from-the-filesystem.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<String> removeSubfolders(String[] folder) {
        Arrays.sort(folder);
        List<String> roots = new ArrayList<>();

        for (String path : folder) {
            if (roots.isEmpty() || !path.startsWith(roots.get(roots.size() - 1) + "/")) {
                roots.add(path);
            }
        }

        return roots;
    }
}

/*
Explanation

After lexicographic sorting, a parent folder appears before its subfolders.
Keep the last accepted root folder; a new path is a subfolder exactly when it
starts with root + "/".

The slash check is critical: "/a/b" is under "/a", but "/ab" is not. Sorting
acts like a lightweight trie because descendants are grouped after parents.

Edge cases: sibling folders with common prefixes; deep nested chains; first
folder is always accepted.

Time complexity: O(n log n * L) for sorting and prefix checks.
Space complexity: O(n) for the result.
*/
